### PR TITLE
[TI] Create and process OTA header

### PR DESCRIPTION
**Pull request by srickardti**
#### Problem

The CC13X2X7_CC26X2X7 OTA implementation does not process the OTA image header.

#### Change overview

* Use the Python script for OAD header creation
* Add a post processing step to generate the Matter OTA image file
* Process the fixed OTA header to find the beginning of the OAD header for update
* Implement the first run function to notify the provider of an applied image
  * First implementation was with a new NV item to store the last run version and compare to the current version
  * Final implementatino uses the OTA state and the target version to notify; if we are applying an update and the target and running versions match, this is the first run

#### Testing

Multiple OTAs with the linux ota-provider-app.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -246,6 +246,7 @@
 	url = https://github.com/TexasInstruments/cc13xx_cc26xx_sdk.git
 	branch = main
 	platforms = cc13x2_26x2
+	ignore = dirty
```
Comment (andy31415) [inline .gitmodules:249]: Why is this required?

Comment (srickardti) [reply .gitmodules:249]: Using the python script now drops a `__pycache__` in the tools directory. I didn't notice it when creating the repo because I had used the compiled executable. I could probably update the repository's .gitignore too.

File: third_party/ti_simplelink_sdk/oad_merge_tool.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2020 Texas Instruments Incorporated
```
Comment (andy31415) [inline third_party/ti_simplelink_sdk/oad_merge_tool.py:1]: Since this is part of the CHIP SDK it should probably have the CHIP license blurb.

Should there be a `#!/usr/bin/env python` to make it executable directly?

File: .gitmodules (hunk 1 above)
Comment (andy31415) [reply .gitmodules:249]: how about adding to the top of the file a `#!/usr/bin/env -S python3 -B` and that should avoid creating of cache files.

our tools/builds should not dirty the source tree.

File: third_party/ti_simplelink_sdk/oad_merge_tool.py (hunk 2 above)
Comment (andy31415) [reply third_party/ti_simplelink_sdk/oad_merge_tool.py:1]: Make that `#!/usr/bin/env -S python -B` instead to avoid cache file creation.

File: examples/all-clusters-app/cc13x2x7_26x2x7/BUILD.gn [hunk 3]
```diff
@@ -26,17 +26,18 @@ import("${ti_simplelink_sdk_build_root}/ti_simplelink_sdk.gni")
 
 assert(current_os == "freertos")
 
+declare_args() {
+  device_vid = "0xFFF1"
```
[the remark this answers is not in the document]
Comment (srickardti) [reply examples/all-clusters-app/cc13x2x7_26x2x7/BUILD.gn:30]: I could add that prefix, but these values are consumed by the application and not the matter configuration files. Then it's used by the TI executable targets. I noticed all other platforms just documented running the `ota_image_tool.py` again with different versions, but that didn't match the version in the persistent storage.

I'll check to see if it can be an int type. This eventually gets placed in a `#define` on the command line so that's my only concern.

File: src/platform/cc13x2_26x2/OTAImageProcessorImpl.cpp [hunk 4]
```diff
@@ -76,10 +81,30 @@ CHIP_ERROR OTAImageProcessorImpl::ProcessBlock(ByteSpan & block)
...
 }
 
+bool OTAImageProcessorImpl::IsFirstImageRun()
+{
+    CHIP_ERROR err;
+    OTARequestorInterface * requestor;
+    uint32_t runningSwVer;
+
+    err = ConfigurationMgr().GetSoftwareVersion(runningSwVer);
+    SuccessOrExit(err);
```
[the remark this answers is not in the document]
Comment (srickardti) [reply src/platform/cc13x2_26x2/OTAImageProcessorImpl.cpp:95]: That is clearer, the goto pattern is habit for me from working on OpenThread.

File: src/platform/cc13x2_26x2/OTAImageProcessorImpl.cpp [hunk 5]
```diff
@@ -131,17 +142,26 @@ static bool writeExtFlashImgPages(NVS_Handle handle, size_t bytesWritten, Mutabl
...
     status = NVS_write(handle, IMG_START + bytesWritten, block.data(), block.size(), NVS_WRITE_POST_VERIFY);
-    return (status == NVS_STATUS_SUCCESS);
+    if (status != NVS_STATUS_SUCCESS)
+    {
+        ChipLogError(SoftwareUpdate, "NVS_write failed status: %d", status);
+        return false;
+    }
+    else
+    {
+        return true;
```
[the remark this answers is not in the document]
Comment (srickardti) [reply src/platform/cc13x2_26x2/OTAImageProcessorImpl.cpp:157]: I'll get that updated

File: third_party/ti_simplelink_sdk/oad_merge_tool.py (hunk 2 above)
Comment (srickardti) [reply third_party/ti_simplelink_sdk/oad_merge_tool.py:1]: I'll update the copy and add the shebang. This isn't the file that creates the cache file, that's in the TI SDK and is run by ninja.

**Review by andy31415 — APPROVED** [review]

**Review by selissia — APPROVED** [review]

File: src/platform/cc13x2_26x2/OTAImageProcessorImpl.cpp [hunk 6]
```diff
@@ -367,8 +400,25 @@ void OTAImageProcessorImpl::HandleProcessBlock(intptr_t context)
...
+        if (imageProcessor->mBlock.size() + imageProcessor->mParams.downloadedBytes < sizeof(imageProcessor->mFixedOtaHeader))
+        {
+            memcpy(header + imageProcessor->mParams.downloadedBytes, imageProcessor->mBlock.data(), imageProcessor->mBlock.size());
+        }
+        else
+        {
+            memcpy(header + imageProcessor->mParams.downloadedBytes, imageProcessor->mBlock.data(),
+                   sizeof(imageProcessor->mFixedOtaHeader) - imageProcessor->mParams.downloadedBytes);
+        }
+    }
```
Comment (carol-apple) [inline src/platform/cc13x2_26x2/OTAImageProcessorImpl.cpp:416]: Are you doing this to remove the header from the block? If so, why not just use `OTAImageHeaderParser::AccumulateAndDecode` to do it for you?

Comment (srickardti) [reply src/platform/cc13x2_26x2/OTAImageProcessorImpl.cpp:416]: Only the fixed header is useful for this platform. It is used to calculate the start of the payload for the device's bootloader. The `OTAImageHeaderParser` doesn't have an API to get the offset of the payload for further processing. I could have added that to the class to allow access to the `mHeaderTlvSize` or to calculate the start of the payload. But that includes a lot of extra code to store and process the TLVs that are not used by this platform.

Comment (carol-apple) [reply src/platform/cc13x2_26x2/OTAImageProcessorImpl.cpp:416]: I wonder if the right thing to do is to split the API (or at least have the option to do either or both) such that you can just get the fixed part without parsing the rest of the TLVs. Not for this PR but would be good to have down the road.

**Review by carol-apple — APPROVED** [review]
